text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,222 @@
+
+Any attribute in this namespace may:
+
+* Hint the compiler to emit a specific error message in a specific situation
+
+Any attribute in this namespace is not allowed to:
+
+* Change the result of the compilation, which means applying such an attribute should never cause an error as long as they are syntactically valid
+
+The compiler is allowed to:
+
+* Ignore the provided hints
+* Use only parts of the provided hints, for example for the proposed `#[diagnostic::on_unimplemented]` only use the `message` option
+* Change this behaviour at any time
+
+The compiler must:
+
+* Verify that the attribute is syntactically correct, which means:
+    + Its one of the accepted attributes
+    + It only contains the allowed options
+    + Any provided option follows the defined syntax
+* Follow the described semantics if the attribute is not ignored.
+
+
+## The `#[diagnostic::on_unimplemented]` attribute
+
+This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. Implementing the later part is optional as outlined in the previous section. Any syntax not explicitly described in this section must be rejected as invalid according to the general rules of the `#[diagnostic]` attribute namespace.
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+
+Each of the options `message`, `label` and `note` are optional. They are separated by comma. The tailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. The error message can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the actual type name.
+
+In addition the `on_unimplemented` attribute provides mechanisms to specify for which exact types a certain message should be emitted via an `if()` option. It accepts a set of filter options. A filter option consists on the generic parameter name from the trait definition and a type path against which the parameter should be checked. These type path could either be a fully qualified path or refer to any type in the current scope. As a special generic parameter name `Self` is added to refer to the `Self` type of the trait implementation. A filter option evaluates to `true` if the corresponding generic parameter in the trait definition matches the specified type. The provided `message`/`note`/`label` options are only emitted if the filter operation evaluates to `true`.
```

Comment:
Is there a reason to change `on` to `if` from `rustc_on_unimplemented`? I don't care either way, I'm just intrigued at the rationale. `on` is already used in other attributes IIRC.

Also, allowing `Self` is something I'd want and is easy to add, but might have pushback from the lang team.

Reply:
I've went for `if` instead of `on` as I also moved the `message`/`notice`/`label` option out of the filter clause. Instead each attribute now has an optional filter in the `if` clause that determines whether a `message`/… should be emitted. I consider the multiple attributes approach to be easier to use in conjunction with something like `#[cfg_attr]` especially around filters containing types behind a feature flag. So `if` feels for fitting for my non-native speaker mind. If `on` is a better fit I'm happy to use `on` instead.